sentry/src/main/java/io/sentry/ServerNameResolvingEventProcessor.java

```diff
@@ -0,0 +1,134 @@
+
+  /** The global hostname cache to speed up localhost hostname resolution. */
+  private final HostnameCache hostnameCache;
+
+  public ServerNameResolvingEventProcessor() {
+    this(new HostnameCache(HOSTNAME_CACHE_DURATION));
+  }
+
+  ServerNameResolvingEventProcessor(final @NotNull HostnameCache hostnameCache) {
+    this.hostnameCache = hostnameCache;
+  }
+
+  @Override
+  public @NotNull SentryEvent process(
+      final @NotNull SentryEvent event, final @Nullable Object hint) {
+    if (event.getServerName() == null) {
+      event.setServerName(hostnameCache.getHostname());
+    }
+    return event;
+  }
+
+  /**
+   * Time sensitive cache in charge of keeping track of the hostname. The {@code
+   * InetAddress.getLocalHost().getCanonicalHostName()} call can be quite expensive and could be
+   * called for the creation of each {@link SentryEvent}. This system will prevent unnecessary costs
+   * by keeping track of the hostname for a period defined during the construction. For performance
+   * purposes, the operation of retrieving the hostname will automatically fail after a period of
+   * time defined by {@link #GET_HOSTNAME_TIMEOUT} without result.
+   */
+  static final class HostnameCache {
+    /** Time before the get hostname operation times out (in ms). */
+    private static final long GET_HOSTNAME_TIMEOUT = TimeUnit.SECONDS.toMillis(1);
+    /** Time for which the cache is kept. */
+    private final long cacheDuration;
+    /** Current value for hostname (might change over time). */
+    @Nullable private volatile String hostname;
+    /** Time at which the cache should expire. */
+    private volatile long expirationTimestamp;
+    /** Whether a cache update thread is currently running or not. */
+    private final @NotNull AtomicBoolean updateRunning = new AtomicBoolean(false);
+
+    private final @NotNull Callable<InetAddress> getLocalhost;
+
+    private HostnameCache(long cacheDuration) {
+      this(cacheDuration, () -> InetAddress.getLocalHost());
+    }
+
+    /**
+     * Sets up a cache for the hostname.
+     *
+     * @param cacheDuration cache duration in milliseconds.
+     * @param getLocalhost a callback to obtain the localhost address - this is mostly here because
+     *     of testability
+     */
+    HostnameCache(long cacheDuration, final @NotNull Callable<InetAddress> getLocalhost) {
+      this.cacheDuration = cacheDuration;
+      this.getLocalhost = Objects.requireNonNull(getLocalhost, "getLocalhost is required");
+    }
+
+    /**
+     * Gets the hostname of the current machine.
+     *
+     * <p>Gets the value from the cache if possible otherwise calls {@link #updateCache()}.
+     *
+     * @return the hostname of the current machine.
+     */
+    private @Nullable String getHostname() {
+      if (expirationTimestamp < System.currentTimeMillis()
+          && updateRunning.compareAndSet(false, true)) {
+        updateCache();
+      }
+
+      return hostname;
```

Comment:
First time we capture an error it just doesn't include a hostname? It scheduled a thread to run and returns `null`?

If so, should we call `updateCache()` to schedule the thread when the class is constructed?

Reply:
True!